### ISSUE TITLE
Some theoritical questions

### ISSUE THREAD

**Issue by ajz34** [fishjojo/pyscfad#1]
Seems an interesting repo :laughing: 

I'm just curious about some theoritical (instead of practial) issues. Not intended to know about current or near future code availibility :joy: 
I'm also not sure that if this repo is a proof-of-existance work, or an efficiency concerned work.

## Is memory control possible for auto-grad?

(nuclear gradient methods, not using orbital gradient to update SCF results in some Newton way, though the latter is also possible)

My knowledge to auto-grad or deep learning is limited, and haven't dived deep into code and theory details. To my knowledge, nuclear gradient could be done by auto-grad if only energy function is given. One can feed a gradient tensor to original tensor manually, so that autograd could be done, even if jax/torch/tf doesn't know how to handle derivative of electronic integrals (like `int1e_ovlp`).

If auto-grad API is not robust enough, one may have to store all eri derivatives into memory, working as the original tensors's attribute (like `some_tensor.grad` in pytorch). Memory cost could be huge for ERIs or RI-3c2e-ERIs; storing all ERIs to disk is also challenging. Is it possible to alleviate this problem?

Writing a memory bounded algorithms for nuclear gradients is a little distressing for me :crying_cat_face: 

## Does MP2 gradient requires first order orbital coefficient derivative?

To my understanding, usual MP2 gradients utilizes Z-vector method to avoid explicitly calculating first orbital coefficient derivative (or usually denoted by U matrix). Does auto-grad scheme also possible utilizing Z-vector method?

A weaker question is that, does SCF gradients requires U matrix in auto-grad scheme? The U matrix could be eliminated by imposing variational conditions (or vir-occ block of Fock matrix in MO basis is zero, so to it's derivatives). But I guess, that is equation derivation work by human, and not easily recognized by auto-grad engine.

**Comment by fishjojo** [comment]
> Is memory control possible for auto-grad?

In principle this is possible. What we can do is to compute `∂F/∂x*dx` batch by batch. However currently, due to the limitations of the AD tools, it is more efficient to compute the full `∂F/∂x` tensor and then contract it with `dx`. I hope this could be improved in the future. But for machine leaning purposes, the systems in the training set are usually small, so the memory cost won't be the biggest problem.
> Does MP2 gradient requires first order orbital coefficient derivative?

The orbital response is captured by the AD of the eigenvalue decomposition, so there is no need to derive it by hand (i.e., no need of solving the CPHF equations).

**Comment by ajz34** [comment]
Thanks for discussion! :laughing: 
Seems eigenvalue decomposition is brilliant :joy_cat:
